File: reverse_gravatar.pl
#!/usr/bin/perl

=head1 Reverse Gravatar

Reverse Gravatar is intended to effectively reverse MD5 hashes to the email
address from which the sum was generated. As the MD5 hash algorithm is
designed to be irreversible, I accomplish this by the classically simple
approach of generating reasonable candidate data to hash and then comparing the
resulting hash against all the known valid hashes. Any match indicates that we
know the data which was hashed.

The candidate data for Gravatar hashes are email addresses, and for current
purposes these are reasonably guessed to be in a limited set of domains, and
optionally include a separator of underscore, period, colon, semicolon, or
comma. These may not always follow the exact allowed characters of the
RFC in question.

=head2 How to pass data to this program

This program reads whitespace separated strings of input from the command line
arguments. If any of these are valid file path descriptors, these are opened
and read for their content. All the remaining arguments, and the
file contents, are then split on whitespace, and either assumed to form name
portions of an email address or, only if they are 32 characters of lowercase
hexadecimal, to be MD5 hashes. When no arguments are specified the program
instead processes standard in as it would a file (as just described).

The program tries all combinations of names, the first character of
names, from 1 to 3 parts with every and no separator, followed by @
and every possible domain (from an internal list) with . and every possible
domain ending (from an internal list).

These lists are:

=head3 Separators

_ . : ; ,

=head3 Domains

gmail hotmail yahoo mailinator aol verizon speakeasy

=head3 Domain endings

com net org edu co.uk fr

=head2 Example usage

reverse_gravatar.pl daniel pastel lamblin fe101e680f0f36bb6082086bbd65444f
reverse_gravatar.pl sampleInput.txt 2>/dev/null

=cut

use warnings;
use strict;
use Digest::MD5 qw/md5_hex/;

my @g=qw//;
my @n=qw//;
my @s=qw/_ . : ;/;
my @d=qw/gmail hotmail yahoo mailinator aol verizon speakeasy/;
my @t=qw/com net org edu co.uk fr/;

# blank seperator and comma
push @s, "", ",";

sub parseArgOrLine {
  shift;
  chomp;
  foreach (split) {
    if (/^[0-9a-f]{32}$/) {
      push @g, $_;
    } else {
      push @n, $_;
    }
  }
}

my @files;
foreach (@ARGV) {
  if (-e $_) {
    push @files, $_;
  } else {
   parseArgOrLine($_);
  }
}

sub parseFiles (@) {
  local @ARGV = @_;
  while(<>) {
    parseArgOrLine($_);
  }
}

if ($#files >= 0 || $#ARGV < 0) {
  parseFiles(@files);
}

print "Searching for: ", join(", ", @g), "\nWith:\n";

# Index hashes
my %g;
foreach (@g) {
  $g{$_} = 1;
}

# name initials and blank name part
my @c=map{substr($_,0,1)} @n;
push @n, @c, "";

# Unique names
my %c;
foreach (@n) {
  $c{$_} = 1;
}
@n = keys %c;

foreach (@n, @s, @d, @t) {
  print "\"$_\"\n";
}

foreach my $fn (@n) {
  if ("" eq $fn) {
    next; # Always minimum 1 name.
  }
  foreach my $mn (@n) {
    foreach my $ln (("" eq $mn) ? "" : @n) {
      foreach my $s (("" eq $mn) ? "" : @s) {
        foreach my $d (@d) {
          foreach my $t (@t) {
            my $e;
            if ($mn ne "") {
              $e = "$fn$s$mn";
            } else {
              $e = "$fn";
            }
            if ($ln ne "") {
              $e .= "$s$ln"
            }
            $e .= "\@$d.$t";
            my $m=md5_hex($e);
            if (exists $g{$m}) {
              print STDOUT "Hash: $m <= Email: $e MATCHES!\n";
            } else {
              print STDERR "Hash: $m <= Email: $e\n";
            }
          }
        }
      }
    }
  }
}

File: ReverseGravatar.java
import java.lang.RuntimeException;
import java.lang.String;
import java.lang.StringBuilder;
import java.lang.System;
import java.io.IOException;
import java.math.BigInteger;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.function.Function;

import javax.xml.bind.DatatypeConverter;

class ReverseGravatar {

  public static final HashSet<String> hashes = new HashSet<>();
  public static final HashSet<String> names = new HashSet<>();
  public static final String[] seperators = {"_", ".", ":", ";", ""};
  public static final String[] domains
      = {"gmail", "hotmail", "yahoo", "mailinator", "aol", "verizon", "speakeasy"};
  public static final String[] tlds = {"com", "net", "org", "edu", "co.uk", "fr"};

  public static void main(String... args) throws NoSuchAlgorithmException, IOException {
    names.add("");
    ArrayList<String> emails = handleArgs(args);
    generateEmails(emails);
    hashAndCheck(emails, getMd5AndEmailViaDatatypeConverter());
  }

  public static ArrayList<String> handleArgs(String[] args) throws IOException {
    for (String arg : args) {
      Path p = Paths.get(arg);
      if (Files.exists(p)) {
        processFile(p);
      } else {
        storeNameOrHash(arg.toLowerCase());
      }
    }
    int n = names.size();
    int combinations = n * n * n * seperators.length * domains.length * tlds.length;
    return new ArrayList<>(combinations);
  }

  public static void processFile(Path p) throws IOException {
    Files.lines(p).flatMap(s -> Arrays.stream(s.toLowerCase().split("\\s+")))
        .forEach(s -> storeNameOrHash(s));
  }

  public static void storeNameOrHash(String s) {
    if (s.matches("[a-f0-9]{32}")) {
      hashes.add(s);
    } else {
      names.add(s);
      names.add(s.substring(0, 1));
    }
  }

  public static void generateEmails(ArrayList<String> emails) {
    HashSet<String> namesOrEmpty = names;
    HashSet<String> empty = new HashSet<>();
    empty.add("");
    for (String fn: names) {
      if (fn.isEmpty()) {
        continue;
      }
      for (String mn: names) {
        namesOrEmpty = mn.isEmpty() ? empty : names;
        for (String ln: namesOrEmpty) {
          for (String s : seperators) {
            for (String d : domains) {
              for (String t : tlds) {
                StringBuilder email = new StringBuilder(fn);
                if (!mn.isEmpty()) {
                  email.append(s).append(mn);
                }
                if (!ln.isEmpty()) {
                  email.append(s).append(ln);
                }
                email.append('@').append(d).append('.').append(t);
                emails.add(email.toString());
              }
            }
          }
        }
      }
    }
  }

  public static void hashAndCheck(ArrayList<String> emails, Function<String, String> md5AndEmail) {
    emails.parallelStream()
        .map(md5AndEmail)
        .forEach(s -> {
          String[] tupple = s.split("≈", 2);
          String hash = tupple[0], email = tupple[1];
          if (hashes.contains(hash)) {
            System.out.printf("Hash: %s <= Email: %s MATCHES!\n", hash, email);
          } else {
            System.err.printf("Hash: %s <= Email: %s\n", hash, email);
          }
        });
  }

  public static Function<String, String> getMd5AndEmailViaDatatypeConverter() {
    return s -> {
      try {
        return DatatypeConverter.printHexBinary(
            MessageDigest.getInstance("MD5").digest(s.getBytes(StandardCharsets.UTF_8)))
                   .toLowerCase()
               + "≈" + s;
      } catch (NoSuchAlgorithmException e) {
        throw new RuntimeException(e);
      }
    };
  }

  public static Function<String, String> getMd5AndEmailViaBigInteger() {
    return s -> {
      try {
        return String.format(
            "%032x≈%s",
            new BigInteger(1,
                           MessageDigest.getInstance("MD5").digest(s.getBytes(StandardCharsets.UTF_8))),
            s);
      } catch (NoSuchAlgorithmException e) {
        throw new RuntimeException(e);
      }
    };
  }
}